### PR TITLE
use ttl handle config to compute creation and expiration timestamps f…

### PR DESCRIPTION
**Pull request by galganif**
…or references when stored with an handle

### CODE REVIEW

File: java/arcs/core/storage/Reference.kt [hunk 1]
```diff
@@ -31,13 +34,25 @@ data class Reference(
     val storageKey: StorageKey,
     val version: VersionMap?,
     /** Reference creation time (in milliseconds). */
-    override val creationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
+    private var creationTimestamp_: Long = RawEntity.UNINITIALIZED_TIMESTAMP,
     /** Reference expiration time (in milliseconds). */
-    override val expirationTimestamp: Long = RawEntity.UNINITIALIZED_TIMESTAMP
+    private var expirationTimestamp_: Long = RawEntity.UNINITIALIZED_TIMESTAMP
```
Comment (csilvestrini) [inline java/arcs/core/storage/Reference.kt:39]: These should use underscore as a prefix, not a suffix. See https://kotlinlang.org/docs/reference/coding-conventions.html#property-names "Names for backing properties" section

File: java/arcs/core/storage/Reference.kt [hunk 2]
```diff
@@ -31,13 +34,25 @@ data class Reference(
...
 ) : Referencable, arcs.core.data.Reference<RawEntity> {
     /* internal */
     var dereferencer: Dereferencer<RawEntity>? = null
 
+    override val creationTimestamp: Long get() = creationTimestamp_
+    override val expirationTimestamp: Long get() = expirationTimestamp_
+
+    fun ensureTimestamps(time: Time, ttl: Ttl) {
```
Comment (csilvestrini) [inline java/arcs/core/storage/Reference.kt:47]: ensure timestamps are... what? This method name is a little strange. Here's a few other suggestions:

ensureTimestampsAreSet
initializeTimestamps
setTimestamps

File: java/arcs/core/entity/StorageAdapter.kt [hunk 3]
```diff
@@ -64,16 +64,22 @@ class EntityStorageAdapter<T : Entity>(
 /** [StorageAdapter] for converting [Reference] to/from [StorageReference]. */
 class ReferenceStorageAdapter<E : Entity>(
     private val entitySpec: EntitySpec<E>,
-    private val dereferencerFactory: EntityDereferencerFactory
+    private val dereferencerFactory: EntityDereferencerFactory,
+    val ttl: Ttl,
+    val time: Time
```
Comment (csilvestrini) [inline java/arcs/core/entity/StorageAdapter.kt:69]: Can these be private?

File: java/arcs/core/entity/Reference.kt [hunk 4]
```diff
@@ -25,12 +27,17 @@ class Reference<T : Entity>(
...
+    val creationTimestamp get() = storageReference.creationTimestamp
+    val expirationTimestamp get() = storageReference.expirationTimestamp
+
     /** Returns the [Entity] pointed to by this reference. */
     suspend fun dereference() = storageReference.dereference()?.let { entitySpec.deserialize(it) }
 
     /** Returns a [Referencable] for this reference. */
     /* internal */ fun toReferencable(): StorageReference = storageReference
 
+    fun ensureTimestamps(time: Time, ttl: Ttl) = storageReference.ensureTimestamps(time, ttl)
```
Comment (csilvestrini) [inline java/arcs/core/entity/Reference.kt:39]: Same comment here as below about the name of this method.

File: java/arcs/core/entity/Reference.kt [hunk 5]
```diff
@@ -25,12 +27,17 @@ class Reference<T : Entity>(
     /** The entity ID for the referenced entity. */
     val entityId = storageReference.id
 
+    val creationTimestamp get() = storageReference.creationTimestamp
+    val expirationTimestamp get() = storageReference.expirationTimestamp
```
Comment (csilvestrini) [inline java/arcs/core/entity/Reference.kt:31]: Insert new line and indent before `get()`

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 6]
```diff
@@ -1231,11 +1237,11 @@ class DatabaseImpl(
...
-            id = it.getString(0),
-            storageKey = StorageKeyParser.parse(it.getString(3)),
-            version = it.getVersionMap(4),
-            creationTimestamp = it.getString(1).toLong(),
-            expirationTimestamp = it.getString(2).toLong()
+            it.getString(0),
+            StorageKeyParser.parse(it.getString(3)),
+            it.getVersionMap(4),
+            it.getString(1).toLong(),
+            it.getString(2).toLong()
```
Comment (csilvestrini) [inline java/arcs/android/storage/database/DatabaseImpl.kt:1244]: I found the named arguments helpful here. You don't really know what they all mean without them; 5 args is a little hard to remember off the top of your head.

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 7]
```diff
@@ -297,11 +297,11 @@ class DatabaseImpl(
...
-                        id = it.getString(6),
-                        storageKey = StorageKeyParser.parse(it.getString(7)),
-                        version = it.getVersionMap(8),
-                        creationTimestamp = it.getLong(9),
-                        expirationTimestamp = it.getLong(10)
+                        it.getString(6),
+                        StorageKeyParser.parse(it.getString(7)),
+                        it.getVersionMap(8),
+                        it.getLong(9),
+                        it.getLong(10)
```
Comment (csilvestrini) [inline java/arcs/android/storage/database/DatabaseImpl.kt:304]: Same comment here as below, about the named args

File: javatests/arcs/core/entity/HandleManagerTestBase.kt [hunk 8]
```diff
@@ -363,6 +363,28 @@ open class HandleManagerTestBase {
         assertThat(handleB.fetch()).isNull()
     }
 
+    @Test
+    fun referenceSingleton_withTTL() = testRunner {
```
Comment (csilvestrini) [inline javatests/arcs/core/entity/HandleManagerTestBase.kt:367]: Nit: capitalisation: `_withTtl()`

**Comment by galganif** [comment]
thanks, all done

**Pull request merged**